v3/app/models/user.rb

```diff
@@ -0,0 +1,114 @@
+class User < ActiveRecord::Base
+
+    has_secure_password
+
+    def self.validate_token(options, check_role_id = 0) # check_role_id default to 0 for 'any'
+      option_token = options[:token].to_s
+      option_ip = options[:ip].to_s
+      option_timenow = Time.now.utc
+      timeok = (option_timenow - 15.minutes).to_s[0,19]
+
+      wheres = sanitize_sql_for_conditions(["ut.token = ? and ut.ip = ?", option_token, option_ip])
+
+      sql = "
+        select ut.*, u.name, u.login, u.role_ids
+        from user_tokens ut
+        inner join users u on u.id = ut.user_id
+        where #{wheres}
+        limit 1
+      "
+      usertoken = (User.find_by_sql sql)[0]
+
+      if !usertoken
+        # INVALID TOKEN OR IP
+        return 400, nil
+      elsif usertoken.timenow.to_s[0,19] < timeok
+        # SESSION TIMEOUT / REMOVE TOKEN
+        sql = "delete from user_tokens ut where #{wheres} limit 1"
+        User.connection.execute sql
+
+        return 401, nil
+      elsif !usertoken.is_role?(check_role_id)
+        # FORBIDDEN / DO NOT RESET USER.TIMENOW
+        return 403, nil
+      else
+        # VALID TOKEN + IP + TIME / RESET USER.TIMENOW
+        usertoken.timenow = option_timenow
+        sql = "update user_tokens ut set timenow = '#{option_timenow.to_s[0,19]}' where #{wheres} limit 1"
+        User.connection.execute sql
+
+        return 200, { :id => usertoken.user_id, :name => usertoken.name, :login => usertoken.login, :role_ids => usertoken.role_ids }
+      end
+    end
+
+    def self.sign_out(options)
+      token = options[:token].to_s
+      ip = options[:ip].to_s
+      all = options[:all]
+
+      wheres = sanitize_sql_for_conditions(["token = ? and ip = ?", token, ip])
+
+      sql = "select * from user_tokens where #{wheres} limit 1"
+      usertoken = (User.find_by_sql sql)[0]
+      return false if !usertoken
+
+      if all
+        sql = "delete from user_tokens where user_id = #{usertoken.user_id}"
+        User.connection.execute sql
+      else
+        sql = "delete from user_tokens ut where #{wheres} limit 1"
+        User.connection.execute sql
+      end
+
+      return true
+    end
+
+    # DOES USER HAVE PERMISSIONS FOR <ROLE_ID>
+    def is_role?(role_id)
```

Comment:
I read this as "is the user a role?" which I know isn't right. maybe `has_role?`

Reply:
Great idea! I like "has_role?"

Reply:
DONE.